src/main/java/com/gtnewhorizons/modularui/common/widget/textfield/NumericWidget.java

```diff
@@ -0,0 +1,387 @@
+package com.gtnewhorizons.modularui.common.widget.textfield;
+
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.text.ParseException;
+import java.util.ArrayList;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.regex.Pattern;
+
+import net.minecraft.network.PacketBuffer;
+
+import com.gtnewhorizon.gtnhlib.util.parsing.MathExpressionParser;
+import com.gtnewhorizon.gtnhlib.util.parsing.MathExpressionParser.Context;
+import com.gtnewhorizons.modularui.ModularUI;
+import com.gtnewhorizons.modularui.api.math.Alignment;
+import com.gtnewhorizons.modularui.api.widget.ISyncedWidget;
+import com.gtnewhorizons.modularui.api.widget.Interactable;
+import com.gtnewhorizons.modularui.config.Config;
+
+/**
+ * A widget that allows the user to enter a numeric value. Synced between client and server. Automatically handles
+ * number parsing and formatting. Only the numeric value (of type <code>double</code>) is exposed to the calling code.
+ * <p>
+ * If GTNHLib is present, also allows entering values as mathematical expressions.
+ */
+public class NumericWidget extends BaseTextFieldWidget implements ISyncedWidget {
+
+    private double value = 0;
+    private Supplier<Double> getter;
+    private Consumer<Double> setter;
+    private Function<Double, Double> validator;
+
+    private double minValue = 0;
+    private double maxValue = Double.POSITIVE_INFINITY;
+    private double defaultValue = 0;
+    private double scrollStep = 1;
+    private double scrollStepCtrl = 0.1;
+    private double scrollStepShift = 100;
+    private boolean integerOnly = true;
+    private NumberFormat numberFormat;
+
+    private Context ctx;
+    private static final int MAX_FRACTION_DIGITS = 4;
+    private static final Pattern NUMBER_PATTERN = Pattern.compile("-?[0-9.,  _’]*");
+    // Character ' ' (non-breaking space) to support French locale thousands separator.
```

Comment:
There should be static constant for this specific character (written using Unicode escape sequence) to prevent confusion in code and unnecessary comments everywhere you handle this character.

Reply:
That would make the pattern declarations look awkward... `Pattern.compile("-?[0-9., " + FRENCH_SPACE + "_’]*");` just reads weird to me and doesn't explain anything more than the comment.

I'm honestly not sure how or even *if* to handle this; the character does not display correctly in game (you get the square symbol thing), so I convert it into a regular space anyway, so the parser realistically never encounters it at all. I'm considering just dropping it altogether and rely on the conversion to spaces.

Reply:
For now I have at least replaced the literal space with the unicode character entity, so that it does not show up completely blank in editors.

Reply:
Regular expression can live without usage of such constant. Most confusion is caused by expressions like `replace(whitespace, whitespace)` - it's not clear which one is which.